// reactstrap-date-picker/lib/props/calendar.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.calendarProps = exports.calendarDefaultProps = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _util = require("./util");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var calendarProps = {
  dayLabels: _propTypes.default.array,
  monthLabels: _propTypes.default.array,
  weekStartsOn: _propTypes.default.number,
  showWeeks: _propTypes.default.bool,
  previousButtonElement: _util.propElemOrString,
  nextButtonElement: _util.propElemOrString,
  pickMonthElement: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.func, _propTypes.default.element]),
  showTodayButton: _propTypes.default.bool,
  todayButtonLabel: _propTypes.default.string,
  roundedCorners: _propTypes.default.bool,
  // better borderRadius
  cellPadding: _propTypes.default.string,
  calendarPlacement: _propTypes.default.oneOfType([_propTypes.default.string, _propTypes.default.func]),
  calendarContainer: _propTypes.default.oneOfType([_propTypes.default.string, _util.propRef])
};
exports.calendarProps = calendarProps;
var calendarDefaultProps = {
  cellPadding: '5px',
  dayLabels: ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'],
  monthLabels: ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December'],
  previousButtonElement: '<',
  nextButtonElement: '>',
  pickMonthElement: undefined,
  calendarPlacement: 'bottom',
  showTodayButton: false,
  todayButtonLabel: 'Today',
  showWeeks: false,
  roundedCorners: false
};
exports.calendarDefaultProps = calendarDefaultProps;
