// reactstrap-date-picker/lib/props/globals.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.globalProps = exports.globalDefaultProps = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _util = require("./util");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var globalProps = {
  value: _propTypes.default.string,
  defaultValue: _propTypes.default.string,
  id: _propTypes.default.string,
  name: _propTypes.default.string,
  dateFormat: _propTypes.default.string,
  // 'MM/DD/YYYY', 'DD/MM/YYYY', 'YYYY/MM/DD', 'DD-MM-YYYY'
  minDate: _propTypes.default.string,
  maxDate: _propTypes.default.string,
  clearButtonElement: _util.propElemOrString,
  showClearButton: _propTypes.default.bool
};
exports.globalProps = globalProps;

var defaultDateFormat = () => {
  var language = typeof window !== 'undefined' && window.navigator ? (window.navigator.userLanguage || window.navigator.language || '').toLowerCase() : '';
  var dateFormat = !language || language === 'en-us' ? 'MM/DD/YYYY' : 'DD/MM/YYYY';
  return dateFormat;
};

var globalDefaultProps = {
  clearButtonElement: '×',
  dateFormat: defaultDateFormat(),
  showClearButton: true
};
exports.globalDefaultProps = globalDefaultProps;
