// reactstrap-date-picker/lib/props/input.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.inputProps = exports.inputDefaultrops = void 0;

var _propTypes = _interopRequireDefault(require("prop-types"));

var _util = require("./util");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

//import React  from 'react'
var inputProps = {
  autoComplete: _propTypes.default.string,
  autoFocus: _propTypes.default.bool,
  disabled: _propTypes.default.bool,
  noValidate: _propTypes.default.bool,
  placeholder: _propTypes.default.string,
  required: _propTypes.default.bool,
  className: _propTypes.default.string,
  style: _propTypes.default.object,
  inputRef: _util.propRef,
  customControl: _propTypes.default.object,
  children: _propTypes.default.oneOfType([_propTypes.default.arrayOf(_propTypes.default.node), _propTypes.default.node])
};
exports.inputProps = inputProps;
var inputDefaultrops = {
  autoFocus: false,
  disabled: false,
  autoComplete: 'on',

  /*style: {
    width: '100%'
  },*/
  noValidate: false
};
exports.inputDefaultrops = inputDefaultrops;
