// reactstrap-date-picker/lib/calendar/CalendarFooter.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CalendarFooter = void 0;

var _react = _interopRequireDefault(require("react"));

var _reactstrap = require("reactstrap");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var CalendarFooter = _ref => {
  var {
    dayLabels,
    showWeeks,
    handleTodayClick,
    showTodayButton,
    todayButtonLabel
  } = _ref;

  if (!showTodayButton) {
    return null;
  }

  return /*#__PURE__*/_react.default.createElement("tfoot", null, /*#__PURE__*/_react.default.createElement("tr", null, /*#__PURE__*/_react.default.createElement("td", {
    colSpan: dayLabels.length + (showWeeks ? 1 : 0),
    style: {
      paddingTop: '9px'
    }
  }, /*#__PURE__*/_react.default.createElement(_reactstrap.Button, {
    block: true,
    size: "sm",
    className: "u-today-button",
    onClick: () => handleTodayClick()
  }, todayButtonLabel))));
};

exports.CalendarFooter = CalendarFooter;
